### PR TITLE
LICM: hoist BO assoc when BinOp is in RHS

### PR DESCRIPTION
**Pull request by artagnon**
Extend hoistBOAssociation smoothly to handle the case when the inner BinaryOperator is in the RHS of the outer BinaryOperator. This completes the generalization of hoistBOAssociation, and the only limitation after this patch is the fact that only Add and Mul are hoisted.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 1]
```diff
@@ -2825,30 +2823,45 @@ static bool hoistBOAssociation(Instruction &I, Loop &L,
...
   auto *Inv = Builder.CreateBinOp(Opcode, C1, C2, "invariant.op");
 
-  auto *NewBO = BinaryOperator::Create(
-      Opcode, LV, Inv, BO->getName() + ".reass", BO->getIterator());
+  auto *NewBO =
+      BinOpInRHS
+          ? BinaryOperator::Create(Opcode, Inv, LV, BO->getName() + ".reass",
+                                   BO->getIterator())
+          : BinaryOperator::Create(Opcode, LV, Inv, BO->getName() + ".reass",
+                                   BO->getIterator());
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/LICM.cpp:2864]: As the operations are commutative, there is no need to retain the original order.

**Review by nikic — APPROVED** [review]
LGTM

A caveat here is that this will not work correctly if both operands are binary operators. So a more general approach would be to instead check whether the LHS is loop invariant and then swap. Unlikely to matter though.

**Comment by artagnon** [comment]
> A caveat here is that this will not work correctly if both operands are binary operators. So a more general approach would be to instead check whether the LHS is loop invariant and then swap. Unlikely to matter though.

Thanks for catching this oversight. I've pushed a fix along with some tests: could you kindly check it?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**